client/src/utils/fileLock.ts

```diff
@@ -0,0 +1,70 @@
+import {open, unlink, readFile} from 'fs/promises';
+
+export class FileLockError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'FileLockError';
+  }
+}
+export async function acquireLock(lockFilePath: string): Promise<boolean> {
+  try {
+    const fileHandle = await open(lockFilePath, 'wx');
+    await fileHandle.writeFile(String(process.pid));
+    await fileHandle.close();
+    return true;
+  } catch (error) {
+    const err = error as NodeJS.ErrnoException;
+    if (err.code === 'EEXIST') {
+      return await handleExistingLock(lockFilePath);
+    }
+    throw new FileLockError(`Failed to acquire lock: ${(error as Error).message}`);
+  }
+}
+
+export async function releaseLock(lockFilePath: string): Promise<void> {
+  try {
+    await unlink(lockFilePath);
+  } catch (error) {
+    const err = error as NodeJS.ErrnoException;
+    if (err.code === 'ENOENT') {
+      return;
+    }
+    throw new FileLockError(`Failed to release lock: ${(error as Error).message}`);
+  }
+}
+
+async function handleExistingLock(lockFilePath: string): Promise<boolean> {
+  try {
+    const pidInLock = await readFile(lockFilePath, 'utf-8');
+    process.kill(Number(pidInLock), 0);
+    return false;
+  } catch (error) {
+    const err = error as NodeJS.ErrnoException;
+    if (err.code === 'ESRCH') {
+      await releaseLock(lockFilePath);
+      return await acquireLock(lockFilePath);
+    }
+    return false;
+  }
+}
+
+const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
+
+export async function lock(lockFilePath: string): Promise<void> {
+  while (!(await acquireLock(lockFilePath))) {
+    await delay(50);
+  }
+}
```

Comment:
무한 대기 방지를 위한 타임아웃 설정이 필요할 것 같아요. 이 부분이 궁금해서 저도 조금 찾아보았는데 '파일 락처럼 공유 자원을 다룰 때는 데드락 방지를 위해 반드시 필요'하다고 하네요!
PID는 존재하지만 실제로는 멈춘 프로세스가 락을 잡고 있는 경우, while(true) 루프로 무한 재시도하거나 락을 못 잡으면 영원히 대기하는 상황이 존재하기 때문에 다음과 같은 개선은 어떨지 제안드려봅니다!

```suggestion
export async function lock(lockFilePath: string, timeoutMs = 30000): Promise<void> {
  const startTime = Date.now();
  
  while (!(await acquireLock(lockFilePath))) {
    if (Date.now() - startTime > timeoutMs) {
      throw new FileLockError(`Lock acquisition timeout after ${timeoutMs}ms`);
    }
    await delay(50);
  }
}
```

Reply:
오~ 너무 좋은 피드백이네요!
저는 handleExistingLock에서 lock은 있지만 pid가 죽었을 때 lock을 해제하고 다시 시도하면 충분할 줄 알았는데 이런 문제가 발생할 수 있군요!

아래 피드백 주신 점과 함께 같이 반영해볼게요!